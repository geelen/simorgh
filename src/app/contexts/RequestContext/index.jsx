import React from 'react';
import { bool, node, oneOf, string, number } from 'prop-types';
import getStatsDestination from './getStatsDestination';
import getStatsPageIdentifier from './getStatsPageIdentifier';
import getOriginContext from './getOriginContext';
import getEnv from './getEnv';
import getMetaUrls from './getMetaUrls';
import variantPropType from '../../models/propTypes/variants';

export const RequestContext = React.createContext({});

export const RequestContextProvider = ({
  children,
  bbcOrigin,
  id,
  isAmp,
  pageType,
  service,
  statusCode,
  previousPath,
  pathname,
  variant,
}) => {
  const { isUK, origin } = getOriginContext(bbcOrigin);
  const env = getEnv(origin);
  const platform = isAmp ? 'amp' : 'canonical';
  const statsDestination = getStatsDestination({
    isUK,
    env,
    service,
  });
  const statsPageIdentifier = getStatsPageIdentifier({
    pageType,
    service,
    id,
  });

  const value = {
    env,
    id,
    isUK,
    origin,
    pageType,
    isAmp,
    platform,
    statsDestination,
    statsPageIdentifier,
    statusCode,
    previousPath,
    variant,
    ...getMetaUrls(origin, pathname),
  };

  return (
    <RequestContext.Provider value={value}>{children}</RequestContext.Provider>
  );
};

RequestContextProvider.propTypes = {
  bbcOrigin: string,
  children: node.isRequired,
  id: string,
  isAmp: bool.isRequired,
  pageType: oneOf([
    'article',
    'frontPage',
    'media',
    'error',
    'MAP',
    'FIX',
    'STY',
    'PGL',
  ]).isRequired,
  service: string.isRequired,
  statusCode: number,
  pathname: string.isRequired,
  previousPath: string,
  variant: variantPropType,
};

RequestContextProvider.defaultProps = {
  bbcOrigin: null,
  id: null,
  statusCode: null,
  previousPath: null,
  variant: null,
};
