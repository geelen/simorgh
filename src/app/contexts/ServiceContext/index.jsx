import React from 'react';
import { node, string } from 'prop-types';
import services from '#lib/config/services/loadableConfig';
import { getVariant } from '../../lib/utilities/variantHandler';
import variantPropType from '../../models/propTypes/variants';

export const ServiceContext = React.createContext({});

export const ServiceContextProvider = ({ children, service, variant }) => {
  const LoadableContextProvider = services[service];

  if (!LoadableContextProvider) {
    return null;
  }

  return (
    <LoadableContextProvider
      Context={ServiceContext}
      dataKey={getVariant({ service, variant })}
    >
      {children}
    </LoadableContextProvider>
  );
};

ServiceContextProvider.propTypes = {
  children: node.isRequired,
  service: string,
  variant: variantPropType,
};

ServiceContextProvider.defaultProps = {
  service: 'default',
  variant: 'default',
};
