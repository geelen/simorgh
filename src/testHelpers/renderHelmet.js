/*
 * Similar to this problem https://github.com/testing-library/react-testing-library/issues/402
 * This helper was created to solve the problem of rendering helmet/head contents in snapshots.
 * Pass in a component that uses helmet somewhere in the component tree.
 * The full html tree is rendered and returned which you can then use to snapshot helmet/head contents.
 * Example of use in a test file:

 it('should render correctly', async () => {
    const html = await renderHelmet(<SomeComponent {...someProps} />);
    expect(html).toMatchSnapshot();
  });

 */

import { render, waitForDomChange } from '@testing-library/react';

export default async component => {
  render(component);

  await waitForDomChange();

  const htmlElement = document.querySelector('html');

  const helmetElements = document.querySelectorAll(
    '[data-react-helmet="true"]',
  );

  const removeHelmetAttributes = el => el.removeAttribute('data-react-helmet'); // remove react-helmet attribute noise from elements

  removeHelmetAttributes(htmlElement); // remove react-helmet attribute noise from elements

  Array.from(helmetElements).forEach(removeHelmetAttributes);

  return document.querySelector('html');
};
