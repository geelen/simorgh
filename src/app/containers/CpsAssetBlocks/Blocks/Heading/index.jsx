import React, { useContext } from 'react';
import { string } from 'prop-types';
import { Headline } from '@bbc/psammead-headings';
import { ServiceContext } from '../../../../contexts/ServiceContext';

const HeadingContainer = ({ idAttr, text }) => {
  const { script, service } = useContext(ServiceContext);

  if (!text) return null;

  return (
    <Headline script={script} service={service} id={idAttr}>
      {text}
    </Headline>
  );
};

HeadingContainer.propTypes = {
  idAttr: string,
  text: string.isRequired,
};

HeadingContainer.defaultProps = {
  idAttr: null,
};

export default HeadingContainer;
