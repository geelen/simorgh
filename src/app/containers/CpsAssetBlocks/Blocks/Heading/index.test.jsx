import React from 'react';
import { shouldMatchSnapshot, isNull } from '@bbc/psammead-test-helpers';
import { ServiceContextProvider } from '../../../../contexts/ServiceContext';
import { suppressPropWarnings } from '../../../../../testHelpers';
import Heading from '.';

describe('CpsAssetBlocks Heading', () => {
  shouldMatchSnapshot(
    'should render correctly',
    <ServiceContextProvider service="news">
      <Heading uuid="uuid" idAttr="idAttr" text="Example text" />
    </ServiceContextProvider>,
  );

  describe('when text isnt provided', () => {
    suppressPropWarnings(['text', 'undefined']);

    isNull(
      'should render null',
      <ServiceContextProvider service="news">
        <Heading uuid="uuid" idAttr="idAttr" />
      </ServiceContextProvider>,
    );
  });
});
