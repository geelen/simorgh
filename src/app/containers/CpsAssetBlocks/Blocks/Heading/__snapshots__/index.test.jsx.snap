// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MediaPageBlocks Heading should render correctly 1`] = `
.c0 {
  font-size: 1.75rem;
  line-height: 2rem;
  font-family: ReithSerif,Helvetica,Arial,sans-serif;
  font-weight: 500;
  font-style: normal;
  color: #3F3F42;
  margin: 0;
  padding: 2rem 0;
}

@media (min-width:20rem) and (max-width:37.4375rem) {
  .c0 {
    font-size: 2rem;
    line-height: 2.25rem;
  }
}

@media (min-width:37.5rem) {
  .c0 {
    font-size: 2.75rem;
    line-height: 3rem;
  }
}

@media (min-width:37.5rem) {
  .c0 {
    padding: 2.5rem 0;
  }
}

<h1
  class="c0"
  id="idAttr"
>
  Example text
</h1>
`;
